### PR TITLE
Backends: pull some macro processing out, add errors

### PR DESCRIPTION
**Pull request by W95Psp**
This PR:
 - move the processing of `secret_bytes` and `secret_arrays` from backends to a proper phase;
 - gives a `mli` signature to backends;
 - flatten a bit backends;
 - adds better error handling facilities for backends.

Makes issue #28 much more green!

@cmester0: I did some changes to the Coq backend, basically I removed a layer of module (which was necessary at some point but now it's not any longer), and by default transformed the TODO nodes into errors, so that it's easier to track what is not actually extracting. Wdyt?

### CODE REVIEW

**Comment by W95Psp** [comment]
I'm thinking a bit about those inline macros: I think we should actually let Rust expand them... Those are generating code we (almost) support.

Expansions contain calls to `vec!`, which produces code we don't support, but I think we should support `vec!` instead of supporting specific hacspeclib macros

**Comment by franziskuskiefer** [comment]
> I'm thinking a bit about those inline macros: I think we should actually let Rust expand them... Those are generating code we (almost) support.
> 
> Expansions contain calls to `vec!`, which produces code we don't support, but I think we should support `vec!` instead of supporting specific hacspeclib macros

Yes! I generally agree. Not expanding them would be a stopgap measure until it's all supported. But if you think it is easy enough to support the code in the macros, we should do that.

File: engine/backends/coq/coq_backend.ml
Comment (cmester0) [inline engine/backends/coq/coq_backend.ml:1]: These changes seems nice. At some point we would probably want to report the errors, and still continue the translation. However this allows you to kinda choose either, so it is a good alternative for now, and reduces confusion.

**Review by cmester0 — APPROVED** [review]

**Comment by W95Psp** [comment]
I've removed all code relative to hacspec-specific inline macros, in favor of their Rust expansion as we discussed

This PR now also adds support for more fundamental rust stuff:
 - (some) pointer expressions, which are actually coercions between internal representations (e.g. sized [u8; N] to unsized slice [u8] or function signatures);
 - repeat expressions, for instance `[123; 18]`;
 - indirectly, this makes `vec!` work properly;
 - translate ZstLiteral nicely, those are actually used (mostly?) for unapplied functions that are specialized with generic types (one could see that as a type-only partial application)

**Pull request merged**